storage2/src/state/transaction.rs

```diff
@@ -0,0 +1,43 @@
+use std::collections::HashMap;
+
+use anyhow::Result;
+
+use jmt::storage::{TreeReader, TreeWriter};
+
+use super::{State, StateWrite};
+
+/// Represents a transactional set of changes to a `State` fork,
+/// implemented as a RYW cache over a `State`.
+pub(crate) struct Transaction<'a, R: TreeReader + TreeWriter> {
+    // TODO: should higher-level types be used here instead of jmt types?
+    cache: HashMap<jmt::KeyHash, jmt::OwnedValue>,
+    unwritten_changes: Vec<(jmt::KeyHash, jmt::OwnedValue)>,
+    state: &'a State<'a, R>,
+}
+
+impl<'a, R: TreeReader + TreeWriter> Transaction<'a, R> {
+    pub fn new() -> Self {
+        Self {
+            cache: todo!(),
+            unwritten_changes: todo!(),
+            state: todo!(),
+        }
+    }
+
+    pub fn commit(self) -> Result<()> {
+        // Write unwritten_changes to the state
+        // `self` will be consumed afterwards
+        todo!()
+    }
```

Comment:
Why does this return a `Result<()>`? Since we're writing changes to the in-memory `State` fork, where does the fallibility come from?

Also, should we be calling this `commit`, if we're not yet writing data to durable storage? Is it potentially confusing to have both `Transaction::commit` and `State::commit`, with different semantics about durability?

Should we call this `Transaction::finish` instead? Or would having two different names be more confusing?

Reply:
> Should we call this Transaction::finish instead? Or would having two different names be more confusing?

I don't have a strong opinion here. Reflexively we're used to "committing" database transactions, but you're right that there are different semantics to this state transaction vs a DB transaction.

Reply:
Cool, let's keep it as `commit` then, I think it's a reasonable choice, we should just be clear that we're committing to our parent in-memory state fork, rather than committing that fork to the storage.